src/algebra/ring/pi.lean

```diff
@@ -0,0 +1,73 @@
+/-
+Copyright (c) 2018 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Simon Hudon, Patrick Massot
+-/
+import tactic.pi_instances algebra.group.pi algebra.ring.basic
+/-!
+# Pi instances for ring
+
+This file defines instances for ring, semiring and related structures on Pi Types
+-/
+
+
+namespace pi
+universes u v w
+variable {I : Type u}     -- The indexing type
+variable {f : I → Type v} -- The family of types already equipped with instances
+variables (x y : Π i, f i) (i : I)
+
+instance mul_zero_class [Π i, mul_zero_class $ f i] : mul_zero_class (Π i : I, f i) :=
+by refine_struct { zero := (0 : Π i, f i), mul := (*), .. }; tactic.pi_instance_derive_field
+
+instance distrib [Π i, distrib $ f i] : distrib (Π i : I, f i) :=
+by refine_struct { add := (+), mul := (*), .. }; tactic.pi_instance_derive_field
+
+instance semiring [∀ i, semiring $ f i] : semiring (Π i : I, f i) :=
+by refine_struct { zero := (0 : Π i, f i), one := 1, add := (+), mul := (*), .. };
+  tactic.pi_instance_derive_field
+
+instance ring [∀ i, ring $ f i] : ring (Π i : I, f i) :=
+by refine_struct { zero := (0 : Π i, f i), one := 1, add := (+), mul := (*),
+  neg := has_neg.neg, .. }; tactic.pi_instance_derive_field
+
+instance comm_ring [∀ i, comm_ring $ f i] : comm_ring (Π i : I, f i) :=
+by refine_struct { zero := (0 : Π i, f i), one := 1, add := (+), mul := (*),
+  neg := has_neg.neg, .. }; tactic.pi_instance_derive_field
+
+/-- A family of ring homomorphisms `f a : γ →+* β a` defines a ring homomorphism
+`pi.ring_hom f : γ →+* Π a, β a` given by `pi.ring_hom f x b = f b x`. -/
+protected def ring_hom
+  {α : Type u} {β : α → Type v} [R : Π a : α, semiring (β a)]
+  {γ : Type w} [semiring γ] (f : Π a : α, γ →+* β a) :
+  γ →+* Π a, β a :=
+{ to_fun := λ x b, f b x,
+  map_add' := λ x y, funext $ λ z, (f z).map_add x y,
+  map_mul' := λ x y, funext $ λ z, (f z).map_mul x y,
+  map_one' := funext $ λ z, (f z).map_one,
+  map_zero' := funext $ λ z, (f z).map_zero }
+
+-- instance is_ring_hom_pi
```

Comment:
Should this comment be removed?

Reply:
Oh, good catch. It looks like this wasn't commented in the original file.